src/sinks/loki.rs

```diff
@@ -0,0 +1,446 @@
+//! Loki sink
+//!
+//! This sink provides downstream support for `Loki` via
+//! the v1 http json endpoint.
+//!
+//! https://github.com/grafana/loki/blob/master/docs/api.md
+//!
+//! This sink does not use `PartitionBatching` but elects to do
+//! stream multiplexing by organizing the streams in the `build_request`
+//! phase. There must be at least one valid set of labels.
+//!
+//! If an event produces no labels, this can happen if the template
+//! does not match, we will add a default label `{agent="vector"}`.
+
+use crate::{
+    dns::Resolver,
+    event::{self, Event, Value},
+    sinks::util::http::{https_client, Auth, BatchedHttpSink, HttpSink},
+    sinks::util::tls::{TlsOptions, TlsSettings},
+    sinks::util::{BatchBytesConfig, TowerRequestConfig, UriSerde},
+    template::Template,
+    topology::config::{DataType, SinkConfig, SinkContext, SinkDescription},
+};
+use derivative::Derivative;
+use futures03::{compat::Future01CompatExt, TryFutureExt};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::collections::HashMap;
+
+type Labels = Vec<(String, String)>;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct LokiConfig {
+    host: UriSerde,
```

Comment:
Are we using `endpoint` now?

Reply:
I wasn't sure if we wanted to be consistent with our current sinks or not since this hasn't been done yet https://github.com/timberio/vector/issues/1775

but I'll go ahead and change this to endpoint :+1: